### PR TITLE
New Relic compatibility

### PR DESCRIPTION
**Pull request by soxjke**
### Abstract

Fixed NewRelic compatibility

### Root cause

NewRelicAgent asserts when _cmd in profiled method is not equal it's name. rac_signalForSelector: introduces "rac_alias_" prefix for swizzled selectors, and it causes NewRelic agent to assert.

### Fix

Synchronously sets back an original implementation for selector while invoking

### Affected areas

3d-party race conditions on runtime calls, performance of method invocation due to additional synchronization against NSObjectRACSignalForSelector() performance gap

### Development testing done

Tested on unit tests and smoke on project used rac_signalForSelector: simultaneously with NewRelicAgent

### CODE REVIEW

File: ReactiveCocoa/NSObject+RACSelectorSignal.m [hunk 1]
```diff
@@ -40,14 +40,19 @@
 @implementation NSObject (RACSelectorSignal)
 
 static BOOL RACForwardInvocation(id self, NSInvocation *invocation) {
-	SEL aliasSelector = RACAliasForSelector(invocation.selector);
+    SEL originalSelector = invocation.selector;
```
Comment (NachoSoto) [inline ReactiveCocoa/NSObject+RACSelectorSignal.m]: Indentation here is wrong.

File: ReactiveCocoa/NSObject+RACSelectorSignal.m [hunk 2]
```diff
@@ -40,14 +40,19 @@
...
 	Class class = object_getClass(invocation.target);
 	BOOL respondsToAlias = [class instancesRespondToSelector:aliasSelector];
 	if (respondsToAlias) {
-		invocation.selector = aliasSelector;
-		[invocation invoke];
+        @synchronized(self) {
+            IMP implementation = method_getImplementation(class_getInstanceMethod(class, aliasSelector));
+            method_setImplementation(class_getInstanceMethod(class, originalSelector), implementation);
+            [invocation invoke];
+            method_setImplementation(class_getInstanceMethod(class, originalSelector), _objc_msgForward);
```
Comment (NachoSoto) [inline ReactiveCocoa/NSObject+RACSelectorSignal.m]: I wonder what the performance implication of this 2 extra calls is.

Comment (mdiep) [inline ReactiveCocoa/NSObject+RACSelectorSignal.m]: Yeah, I think there should be some benchmarking to understand the performance implications before merging.

**Comment by mdiep** [comment]
> NewRelicAgent asserts when cmd in profiled method is not equal it's name. rac_signalForSelector: introduces "rac_alias" prefix for swizzled selectors, and it causes NewRelic agent to assert.

Maybe New Relic should remove this assert?

**Comment by mdiep** [comment]
Closing this because the feedback hasn't been addressed.

**Pull request closed without merging**